java/ql/lib/semmle/code/java/frameworks/spring/SpringWebClient.qll

```diff
@@ -27,3 +27,66 @@ class SpringWebClient extends Interface {
     this.hasQualifiedName("org.springframework.web.reactive.function.client", "WebClient")
   }
 }
+
+private import semmle.code.java.security.RequestForgery
+
+private class SpringWebClientRestTemplateGetForObject extends RequestForgerySink {
+  SpringWebClientRestTemplateGetForObject() {
+    exists(Method m, MethodCall mc, int i |
+      m.getDeclaringType() instanceof SpringRestTemplate and
+      m.hasName("getForObject") and
+      mc.getMethod() = m and
+      // Note that mc.getArgument(0) is modeled separately. This model is for
+      // arguments beyond the first two. There are two relevant overloads, one
+      // with third parameter type `Object...` and one with third parameter
+      // type `Map<String, ?>`. For the latter we cannot deal with mapvalue
+      // content easily but there is a default implicit taint read at sinks
+      // that will catch it.
+      this.asExpr() = mc.getArgument(i + 2) and
+      i >= 0
+    |
+      // If we can determine that part of mc.getArgument(0) is a hostname
+      // sanitizing prefix, then we count how many placeholders occur before it
+      // and only consider that many arguments beyond the first two as sinks.
+      // For the `Map<String, ?>` overload this has the effect of only
+      // considering the map values as sinks if there is at least one
+      // placeholder in the URL before the hostname sanitizing prefix.
+      exists(HostnameSanitizingPrefix hsp |
+        hsp = mc.getArgument(0) and
+        i <=
+          max(int occurrenceIndex, int occurrenceOffset |
+            exists(
+              hsp.getStringValue().regexpFind("\\{[^}]*\\}", occurrenceIndex, occurrenceOffset)
```

Comment:
I'm not immediately familiar, but is there an escaping for literal `{` in these strings that ought to be checked and disregarded?

Reply:
The strings will be something like "http://{domain}.com/users/{userid}". Depending on which overload of `getForObject` you are calling the placeholders are then replaced with either (a) the arguments corresponding to the varargs parameter, in order (ignoring the string in the placeholder), or (b) the value in the provided map, where the key is the string in the placeholder.

In this particular part of the code we're trying to find how many placeholders there are before the "/" at the end of the host name. I guess it would be possible that someone would put an escaped "{" in there. I'm not sure how it would be interpreted. I guess we could first replaceall "\{" with some other string with the same number of characters, say "  ". I'll do that.